### PR TITLE
Add version switcher

### LINKED ISSUE #58190: Add button to switch between std docs version on doc.rust-lang.org

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #58190.

screenshot 2019-02-20 at 17 50 52

r? @QuietMisdreavus

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
cc https://github.com/rust-lang/rust/issues/44687 @rust-lang/docs

**Comment by GuillaumeGomez** [comment]
It's not complete yet. I'm not really satisfied on how I handle the raw HTML. I think that I'll create "temporary" files instead and reference to them.

**Comment by GuillaumeGomez** [comment]
It's now ready for review.

**Comment by GuillaumeGomez** [comment]
Finally fixed! \o/

**Comment by GuillaumeGomez** [comment]
ping @rust-lang/docs

File: src/bootstrap/doc.rs
Comment (frewsxcv) [inline src/bootstrap/doc.rs]: would be cool if we could also add bugfix releases in here, but that might be hard

**Review by frewsxcv — APPROVED** [review]
🥇

File: src/bootstrap/doc.rs
Comment (GuillaumeGomez) [reply src/bootstrap/doc.rs]: Bugfixes don't change API normally so I'm not sure it's really useful to add them in here.

**Comment by GuillaumeGomez** [comment]
ping @frewsxcv

**Comment by frewsxcv** [comment]
I haven't approved a rustdoc PR before and am not familiar with the source so I don't feel comfortable approving this

**Comment by QuietMisdreavus** [comment]
I'll need to give this PR an actual read-through, but i wanted to make sure that this didn't derail the discussion that was happening on https://github.com/rust-lang/rust/issues/44687 last year. I don't remember the specifics of that discussion (it's been a while), but this seems like a different approach than the people in that thread were considering. If it achieves the same goals, then we can fold it into that effort. But this came up independently, so i want to make sure it's not getting in the way.

**Comment by GuillaumeGomez** [comment]
From my point of view, it doesn't prevent the #44687 to happen if they think it's not solving the issue but I'm always glad to wait for reviews coming from you. :)

File: src/librustdoc/lib.rs
Comment (QuietMisdreavus) [inline src/librustdoc/lib.rs]: Bikeshed warning: Since the included file eventually gets named `additional.js`/`additional.css`, i feel like these flags would be better named `--additional-js`/`--additional-css`.

I also feel like the flag description is a little misleading. Maybe something like:

> (JS|CSS) file to include as a static file to be linked from generated documentation

(I feel like this feature would be much better served if it allowed multiple files, but that can be added in a separate PR. Since the flag will be unstable, we can edit it after it lands.)

**Comment by ollie27** [comment]
I would've expected the version selector to be injected by the server running on doc.rust-lang.org because it needs to be injected into the older docs as well as all of the books. It will also need to be kept up to date with the list of released stable versions.

**Comment by GuillaumeGomez** [comment]
The files are just generated then copied into the server. There is nothing dynamic in there. The only thing we can do is injecting the JS/CSS content required to have it.

**Comment by Centril** [comment]
ping from triage @rust-lang/docs

**Comment by jonas-schievink** [comment]
Nominating to move this forward

**Comment by GuillaumeGomez** [comment]
I don't think it'll. It needs to be added directly to the website so it has nothing to do in rustdoc directly. Closing it.

**Pull request closed without merging**
